src/coap_mbedtls.c

```diff
@@ -0,0 +1,1784 @@
+  coap_session_t *c_session = (struct coap_session_t *)ctx;
+
+  if (c_session) {
+    coap_mbedtls_env_t *m_env = (coap_mbedtls_env_t *)c_session->tls;
+    result = coap_session_send(c_session, send_buffer, send_buffer_length);
+    if (result != (ssize_t)send_buffer_length) {
+      coap_log(LOG_WARNING, "coap_network_send failed (%zd != %zd)\n",
+               result, send_buffer_length);
+      result = 0;
+    }
+    else if (m_env) {
+      coap_tick_t now;
+      coap_ticks(&now);
+      m_env->last_timeout = now;
+    }
+  } else {
+    result = 0;
+  }
+  return result;
+}
+
+#if defined(MBEDTLS_SSL_PROTO_DTLS) && defined(MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED) && defined(MBEDTLS_SSL_SRV_C)
+/*
+ * Server side PSK callback
+ */
+static int psk_server_callback(void *p_info, mbedtls_ssl_context *ssl,
+                  const unsigned char *name, size_t name_len )
+{
+  coap_session_t *c_session = (coap_session_t *)p_info;
+  uint8_t buf[128];
+  size_t psk_len;
+  coap_dtls_spsk_t *setup_data;
+  coap_mbedtls_env_t *m_env;
+
+  coap_log(LOG_DEBUG, "got psk_identity: '%.*s'\n",
+                      (int)name_len, name);
+
+  if (c_session == NULL || c_session->context == NULL ||
+    c_session->context->get_server_psk == NULL) {
+    return -1;
+  }
+  m_env = (coap_mbedtls_env_t *)c_session->tls;
+  setup_data = &c_session->context->spsk_setup_data;
+
+  if (setup_data->validate_id_call_back) {
+    coap_bin_const_t lidentity;
+    lidentity.length = name_len;
+    lidentity.s = (const uint8_t*)name;
+    const coap_bin_const_t *psk_key =
+             setup_data->validate_id_call_back(&lidentity,
+                                               c_session,
+                                               setup_data->id_call_back_arg);
+
+    if (psk_key == NULL)
+      return -1;
+    mbedtls_ssl_set_hs_psk(ssl, psk_key->s, psk_key->length);
+    coap_session_refresh_psk_key(c_session, psk_key);
+    m_env->seen_client_hello = 1;
+    return 0;
+  }
+
+  psk_len = c_session->context->get_server_psk(c_session,
+                               (const uint8_t*)name,
+                               name_len,
+                               (uint8_t*)buf, sizeof(buf));
+  m_env->seen_client_hello = 1;
+  mbedtls_ssl_set_hs_psk(ssl, buf, psk_len);
+  return 0;
+}
+#endif /* MBEDTLS_SSL_PROTO_DTLS && MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED && MBEDTLS_SSL_SRV_C */
+
+static char*
+get_san_or_cn_from_cert(mbedtls_x509_crt *crt)
+{
+  if (crt) {
+    mbedtls_asn1_named_data * cn_data;
+
+    if (crt->ext_types & MBEDTLS_X509_EXT_SUBJECT_ALT_NAME) {
+      mbedtls_asn1_sequence *seq = &crt->subject_alt_names;
+      while (seq && seq->buf.p == NULL) {
+        seq = seq->next;
+      }
+      if (seq) {
+        /* Return the Subject Alt Name */
+        return mbedtls_strndup((const char *)seq->buf.p,
+                             seq->buf.len);
+      }
+    }
+
+    /* MBEDTLS_OID_AT_CN is actually only 3 bytes long, but sizeof() gives 4 */
```

Comment:
`MBEDTLS_OID_SIZE(MBEDTLS_OID_AT_CN)` does the trick.

Reply:
Suggestion used.